### PR TITLE
Guild Templates

### LINKED ISSUE #33: guild templates
https://discord.com/developers/docs/resources/guild-template

### PR DESCRIPTION
**Pull request by EveCodes31**
Closes #33

### CODE REVIEW

File: api/template.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+import "time"
+
+// GuildTemplate is a template used for copying guilds https://discord.com/developers/docs/resources/guild-template
+type GuildTemplate struct {
+	Code          string    `json:"code"`
+	Name          string    `json:"name"`
+	Description   *string   `json:"description,omitempty"`
+	UsageCount    int       `json:"usage_count"`
+	CreatorID     Snowflake `json:"creator_id"`
+	Creator       User      `json:"creator"`
```
Comment (TopiSenpai) [inline api/template.go:12]: tbh this should be thrown into the EntityBuilder
```suggestion
	Creator       *User     `json:"creator"`
```

File: internal/disgo_impl.go [hunk 2]
```diff
@@ -281,3 +281,13 @@ func (d *DisgoImpl) SetGuildCommandsPermissions(guildID api.Snowflake, commandPe
...
 	return d.RestClient().SetGuildCommandPermissions(d.ApplicationID(), guildID, commandID, permissions)
 }
+
+// GetTemplate gets a api.GuildTemplate by it's code
+func (d *DisgoImpl) GetTemplate(code string) (*api.GuildTemplate, restclient.RestError) {
+	return d.RestClient().GetGuildTemplate(code)
+}
+
+// CreateGuildFromTemplate creates an api.Guild using a api.Template code
+func (d *DisgoImpl) CreateGuildFromTemplate(templateCode string, data api.CreateGuildFromTemplate) (*api.Guild, restclient.RestError) {
```
Comment (TopiSenpai) [inline internal/disgo_impl.go:291]:
```suggestion
func (d *DisgoImpl) CreateGuildFromTemplate(templateCode string, createGuildFromTemplate api.CreateGuildFromTemplate) (*api.Guild, restclient.RestError) {
```

File: internal/restclient_impl.go [hunk 3]
```diff
@@ -896,6 +896,87 @@ func (r *restClientImpl) DeleteFollowupMessage(applicationID api.Snowflake, inte
...
+	if err != nil {
+		return nil, restclient.NewError(nil, err)
+	}
+
+	rErr = r.Do(compiledRoute, nil, &guildTemplates)
+
+	return
+}
+
+func (r *restClientImpl) CreateGuildTemplate(guildID api.Snowflake, createTemplate api.CreateGuildTemplate) (guildTemplate *api.GuildTemplate, rErr restclient.RestError) {
```
Comment (TopiSenpai) [inline internal/restclient_impl.go:921]:
```suggestion
func (r *restClientImpl) CreateGuildTemplate(guildID api.Snowflake, createGuildTemplate api.CreateGuildTemplate) (guildTemplate *api.GuildTemplate, rErr restclient.RestError) {
```

File: api/disgo.go [hunk 4]
```diff
@@ -52,6 +52,9 @@ type Disgo interface {
 	GetGuildCommandPermissions(guildID Snowflake, commandID Snowflake) (*GuildCommandPermissions, restclient.RestError)
 	SetGuildCommandsPermissions(guildID Snowflake, commandPermissions ...SetGuildCommandPermissions) ([]*GuildCommandPermissions, restclient.RestError)
 	SetGuildCommandPermissions(guildID Snowflake, commandID Snowflake, permissions SetGuildCommandPermissions) (*GuildCommandPermissions, restclient.RestError)
+
+	GetTemplate(code string) (*GuildTemplate, restclient.RestError)
+	CreateGuildFromTemplate(templateCode string, data CreateGuildFromTemplate) (*Guild, restclient.RestError)
```
Comment (TopiSenpai) [inline api/disgo.go:57]:
```suggestion
	CreateGuildFromTemplate(templateCode string, createGuildFromTemplate CreateGuildFromTemplate) (*Guild, restclient.RestError)
```

File: api/restclient.go [hunk 5]
```diff
@@ -92,4 +92,12 @@ type RestClient interface {
 	SendFollowupMessage(applicationID Snowflake, interactionToken string, messageCreate MessageCreate) (*Message, restclient.RestError)
 	UpdateFollowupMessage(applicationID Snowflake, interactionToken string, messageID Snowflake, messageUpdate MessageUpdate) (*Message, restclient.RestError)
 	DeleteFollowupMessage(applicationID Snowflake, interactionToken string, followupMessageID Snowflake) restclient.RestError
+
+	GetGuildTemplate(templateCode string) (*GuildTemplate, restclient.RestError)
+	GetGuildTemplates(guildID Snowflake) ([]*GuildTemplate, restclient.RestError)
+	CreateGuildTemplate(guildID Snowflake, data CreateGuildTemplate) (*GuildTemplate, restclient.RestError)
```
Comment (TopiSenpai) [inline api/restclient.go:98]:
```suggestion
	CreateGuildTemplate(guildID Snowflake, createGuildTemplate CreateGuildTemplate) (*GuildTemplate, restclient.RestError)
```

**Review by TopiSenpai — CHANGES_REQUESTED** [review]
basically looks looks alright except some picky naming & return type

I can do the EntityBuilder part :)

File: api/disgo.go (hunk 4 above)
Comment (TopiSenpai) [reply api/disgo.go]: this one & a few others are still missing 🤔

Comment (EveCodes31) [reply api/disgo.go]: it doesn't appear to be missing on the latest commits?

**Comment by TopiSenpai** [comment]
did the entity builder changes

**Comment by TopiSenpai** [comment]
why are checks not working? 🤔

**Comment by EveCodes31** [comment]
looks like checks are only done on commits to dev?

**Comment by TopiSenpai** [comment]
ahhh because this is a pr to another branch
should prob change that

**Comment by EveCodes31** [comment]
yeah, just make it general

**Pull request merged**